_posts/2018-10-01-safety-musings.md

```diff
@@ -0,0 +1,75 @@
+---
+author: Jorge Gueorguiev Garcia
+layout: post
+asset-type: post
+title: "Safety Musings"
+date: 2018-09-29 08:00:00
+description: Where I muse about safety
+image: 
+    src: /assets/custom/img/blog/2018-09-03-rest-review/relaxfortwo.jpg
+tags: 
+- Software Craftsmanship
+---
+
+So here I am, again, this time with some serious musing around what we do as engineers/developers/craftpeople. Our languages, tools, and techniques, determine what we can do, how good our systems are, how safe, how secure, and how fit for the task. Because this affects people's lives we need to be careful on what we do. Robert Martin (Unble Bob) has been talking about the necessity of a code similar to the Hippocratic oath. I am of the believe that it will come the time in which we will need to be part of a professional body to perform our duties (after all, the Hyppocratic oath did not stop the need for the Medical license)
+
+## Discussing about what are functional languages
+
+The other night I was discussing with some colleagues what is a Functional Language. On their definition, only pure statically typed languages can be called Functional Languages. No IO Monad, no functional language. My contention was that immutability (and with it the reduction of side effects) is enough for a language to be called functional. Their thinking eliminates Clojure, Elixir, Erlang and a few others. Instead of seeing functional languages as a spectrum, it was an all or nothing.
```

Comment:
I assume that you are referring to our chat the other day. I would like to clarify something that maybe I didn't communicate correctly. 

I don't necessarily meant that you need an IO monad to do functional programming, but a way of controlling side effects with the type system. IO monad is a way of doing it and probably the most popular, but not the only one

Reply:
Clojure controls side effects but not through the type system, it does it simply by lacking an assignment operator.

Personally I think the strong vs. dynamic typing spectrum is totally orthogonal to the imperative vs. functional spectrum.

Reply:
It was indeed all steaming from our conversation. I will include your addendum. But to be honest, the thing is that because of that way of thinking I started to ponder about it, and in parts actually agree with you, and from that I evolved the whole post :-)

Reply:
I do agree with this sentence:

"Personally I think the strong vs. dynamic typing spectrum is totally orthogonal to the imperative vs. functional spectrum."

However I also believe the following is true: 

"One doesn't really have control of side effects unless one is constraint from calling inpure code from pure code. Once you invoke an effectful function from a pure function, any callee of it and its ancestors become inpure." 

Honestly I don't know how you can do that without types, but if possible then I would consider that as valid as a typed solution. 

The sentence above have a lot of implications in how you write and arquitect FP apps in the large. It also summarizes how I believe FP code should be written and even though this is very opinionated, any code that doesn't follow that I wouldn't consider fully functional.

We can argue about being pragmatic about the rule, which probably applies here in a way, but still then we would be losing the ability to reason about effects in a safe manner in that part of the code. Whether you need/want to do that in a particular application can also be argued.

My view goes more in the direction of being able to control side effects and needing types to do that, than considering static/dynamic code functional/not functional.

BTW I wish we had more chats like this more often, I enjoy this :)

Reply:
Maybe an example would help.

Taking the above to the exteme, printLn has side effects, therefore any code that calls printLn and all its ancestors now have side effects and are therefore not pure.

Reply:
PrintLn is the less interesting of the side effects :-) Something like a web call or a database query are more interesting to show as an example.

I have expanded the idea of types, as they are four kind of types available to languages (check the updated blog).